### ISSUE TITLE
ToListAsync() thread problems with LoadWith() queries

### ISSUE THREAD

**Issue by cal-tlabwest** [linq2db/linq2db#4404]
`ToListAsync()` has a thread problem when called with a non `ExpressionQuery<>` source such as `LoadWithQueryable<>`.
`db.MyTable.Where(...).LoadWith(...).ToListAsync()` can crash as the DB connection is used in a thread pool task.
`db.MyTable.LoadWith(...).Where(...).ToListAsync()` works however as the source is a `ExpressionQuery<>` in that case.

```
You can get various thread problem related crashes, such as:

Cannot access a disposed object.
Object name: 'SQLiteCommand'.
at System.Data.SQLite.SQLiteCommand.set_Transaction(SQLiteTransaction value)
   at LinqToDB.Data.DataConnection.<>c.<BeginTransaction>b__181_1(DataConnection dataConnection, IsolationLevel isolationLevel)
   at LinqToDB.Data.DataConnection.TraceAction[TContext,TResult](DataConnection dataConnection, TraceOperation traceOperation, Func`2 commandText, TContext context, Func`3 action)
   at LinqToDB.Data.DataConnection.BeginTransaction(IsolationLevel isolationLevel)
   at LinqToDB.Linq.ExpressionQuery`1.System.Collections.Generic.IEnumerable<T>.GetEnumerator()
   at LinqToDB.LinqExtensions.LoadWithQueryable`2.GetEnumerator()
   at System.Linq.Enumerable.TakeWhileIterator[TSource](IEnumerable`1 source, Func`2 predicate)+MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at LinqToDB.AsyncExtensions.<>c__DisplayClass8_0`1.<ToListAsync>b__0()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
   at LinqToDB.AsyncExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken token)

or

Cannot set Transaction while a DataReader is active
 at System.Data.SQLite.SQLiteCommand.set_Transaction(SQLiteTransaction value)
   at LinqToDB.Data.DataConnection.<>c.<BeginTransaction>b__181_1(DataConnection dataConnection, IsolationLevel isolationLevel)
   at LinqToDB.Data.DataConnection.TraceAction[TContext,TResult](DataConnection dataConnection, TraceOperation traceOperation, Func`2 commandText, TContext context, Func`3 action)
   at LinqToDB.Data.DataConnection.BeginTransaction(IsolationLevel isolationLevel)
   at LinqToDB.Linq.ExpressionQuery`1.System.Collections.Generic.IEnumerable<T>.GetEnumerator()
   at LinqToDB.LinqExtensions.LoadWithQueryable`2.GetEnumerator()
   at System.Linq.Enumerable.TakeWhileIterator[TSource](IEnumerable`1 source, Func`2 predicate)+MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at LinqToDB.AsyncExtensions.<>c__DisplayClass8_0`1.<ToListAsync>b__0()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
   at LinqToDB.AsyncExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken token)

or

bad parameter or other API misuse
 at System.Data.SQLite.SQLite3.Bind_DateTime(SQLiteStatement stmt, SQLiteConnectionFlags flags, Int32 index, DateTime dt)
   at System.Data.SQLite.SQLiteStatement.BindParameter(Int32 index, SQLiteParameter param)
   at System.Data.SQLite.SQLiteCommand.BuildNextCommand()
   at System.Data.SQLite.SQLiteDataReader.NextResult()
   at System.Data.SQLite.SQLiteDataReader..ctor(SQLiteCommand cmd, CommandBehavior behave)
   at System.Data.SQLite.SQLiteCommand.ExecuteReader(CommandBehavior behavior)
   at LinqToDB.Data.DataConnection.ExecuteReader(CommandBehavior commandBehavior)
   at LinqToDB.Data.DataConnection.ExecuteDataReader(CommandBehavior commandBehavior)
   at LinqToDB.Linq.QueryRunner.ExecuteElement[T](Query query, IDataContext dataContext, Mapper`1 mapper, Expression expression, Object[] ps, Object[] preambles)
   at LinqToDB.Linq.ExpressionQuery`1.System.Linq.IQueryProvider.Execute[TResult](Expression expression)

```

### Steps to reproduce
```cs
var task = db.MyTable.Where(...).LoadWith(...).ToListAsync();
db.MyTable.Where(...).ToList();  // Will cause thread problems in 1 of 10 runs as the db connection is used from two threads
await task;
```

### Environment details

Linq To DB version: 5.3.2

Database (with version): SQLite 3.42.0

ADO.NET Provider (with version): System.Data.SQLite 1.0.118.0
Operating system: Win11 x64

.NET Version: 8.0

**Comment by jods4** [comment]
This is a problem in your code.
`DbConnection` (ADO.NET) is not thread-safe. You cannot run a query (first `ToListAsync()` call) and then, in parallel, start a second query on the same `DbConnection` (second `ToList()`).
You must await the first query before starting the second, or use two `DbConnection`s if you want to perform two DB queries in parallel.

**Comment by cal-tlabwest** [comment]
No, please consider async workflows. The program flow will continue at the async root even if I await the task directly. In a GUI application it might be a click event that calls ToListAsync() and while this task is executed on the thread pool the GUI thread will continue running the GUI message loop/dispatcher - await will never block the calling thread, it will just continue execution at the async root.
Similar scenarios exists in the ASP.NET Core / .NET generic host - await will continue running code in the calling thread. We can not assume that the calling thread will not make more DB calls during the await.
I guess the solution would be for ToListAsync() to be more like CountAsync() that checks for IQueryProviderAsync instead of ExpressionQuery<> to avoid the specific problem with LoadWith().

**Comment by viceroypenguin** [comment]
@cal-tlabwest: as @jods4 pointed out, the problem is that ado.net is not thread-safe. efc, dapper, and l2db _all_ rely on ado.net, and are explicitly non-thread-safe for a single connection, due to this. if you want to run multiple threads, you must use multiple connections. it is fundamental to how interacting with databases work in C#.

**Comment by jods4** [comment]
This is not a linq2db thing. ADO.NET is not thread-safe, you cannot use a single `DbConnection` on multiple threads like that.
If you to run 2 commands on 2 threads at the same time, you'll need 2 `DbConnection`s.

linq2db `DataConnection` is an abstraction on top of `DbConnection` so it can't be thread-safe either.

**Comment by kaleidocore** [comment]
@jods4 @viceroypenguin I believe you are missing the point here, @cal-tlabwest is not asking to run multiple query threads on the same context but the very opposite - he's pointing out that Linq2db is spawning new threads inside `ToListAsync()` by incorrectly dumping the async query on `GetTask(...)`. The problem seems to be that `ToListAsync()` does not check for `IQueryProviderAsync`, which is why the expression order of `table.LoadWith(...).Where(...).ToListAsync()` works, while the same `table.Where(...).LoadWith(...).ToListAsync()` does not.
Consider the source of `ToListAsync()`:

```
		public static async Task<List<TSource>> ToListAsync<TSource>(
			this IQueryable<TSource> source,
			CancellationToken token = default)
		{
			if (source is ExpressionQuery<TSource> query)
			{
				var list = new List<TSource>();
				await query.GetForEachAsync(list.Add, token).ConfigureAwait(Common.Configuration.ContinueOnCapturedContext);
				return list;
			}

			if (LinqExtensions.ExtensionsAdapter != null)
				return await LinqExtensions.ExtensionsAdapter.ToListAsync(source, token).ConfigureAwait(Common.Configuration.ContinueOnCapturedContext);

			return await GetTask(() => source.AsEnumerable().TakeWhile(_ => !token.IsCancellationRequested).ToList(), token).ConfigureAwait(Common.Configuration.ContinueOnCapturedContext);
		}
```

In comparison, CountAsync() works because it delegates to IQueryProviderAsync:

```
		public static Task<int> CountAsync<TSource>(
			this IQueryable<TSource> source,
			CancellationToken token = default)
		{
			var provider = source.Provider as IQueryProviderAsync;

			if (provider != null)
			{
				return provider.ExecuteAsync<int>(
					Expression.Call(
						null,
						MethodHelper.GetMethodInfo(new Func<IQueryable<TSource>,int>(Queryable.Count), source),
						source.Expression) as Expression,
					token);
			}

			if (LinqExtensions.ExtensionsAdapter != null)
				return LinqExtensions.ExtensionsAdapter.CountAsync<TSource>(source, token);

			return GetTask<int>(source.Count, token);
		}

```

It seems unreasonable to me that using the Linq2db API requires intricate knowledge of internal implementation details in order to stack the `.LoadWith() `and `.Where()` expressions in a particular order, and to know that CountAsync() works while ToListAsync() sometimes crashes.

**Comment by viceroypenguin** [comment]
@kaleidocore This code is, explicitly, invalid with all db providers _and_ orms:
```cs
var task = db.MyTable.Where(...).LoadWith(...).ToListAsync();
db.MyTable.Where(...).ToList();
await task;
```

The internal mechanisms of l2db are not relevant to state that this is invalid code, as it relies on running l2db, and more specifically, ado.net, on two different threads. That it _may_, accidentally, run correctly in some cases is not proof that it _should_ work in all cases. 

Due to the behavior of the db connection between C# and all current db engines, using `DbConnection`, it is not valid behavior to run a query inside of an `async` task, and simultaneously run a query synchronously. Similarly, it is not valid behavior to two `async` tasks in parallel on the same db connection. 

The above code **must** be corrected to either `await task` before executing another db call, or use a second db connection to handle the sync call. Any other discussion about behavior for thread-safety is not relevant and should not be entertained.

Again, this is not just due to l2db construction, this is due to ado.net api, and more specifically, due to the fact that most db connection protocols do not support multiplexing (MARS is not the same thing as multiplexing).

DB Connections must be independent per simultaneously operating threads.

**Comment by kaleidocore** [comment]
@viceroypenguin Very interesting, I appreciate the information. So, just to make sure I've understood this correctly, this (pseudo-) code would be explicitly forbidden:

```
class MyForm : Form
{
  DbContext _db;
  int _count;

  async Task HandleButtonClick()
  {
    _count = await _db.MyTable.CountAsync();
  }
}

```
In other words, a DbContext may never live beyond the scope of the function it is defined in, if it at any point is involved in an async operation?

**Comment by viceroypenguin** [comment]
@kaleidocore - that would be fine. A DbContext can live beyond an async method. It is not about the lifetime. It is about multi-threaded behavior. You can do multiple asyncs and share dbcontext across multiple methods. You just cannot share a single instance across multiple threads simultaneously. For example, you could not do:

```cs
var task1 = CountAsync();
var task2 = ToListAsync();

await Task.WhenAll(task1, task2);
```

But you could do:

```cs
var count = await CountAsync();
var list = await ToListAsync();
```

In this case, the db context is not being shared simultaneously across multiple threads.

**Comment by kaleidocore** [comment]
@viceroypenguin But that click handler is invoked again if you click twice - and from what I find `ToListAsync()` spawns another concurrent thread through GetTask(...) if invoked with `table.Where(...).LoadWith(...).ToListAsync()` but not when invoked it with `table.LoadWith(...).Where(...).ToListAsync()`. 

```
class MyForm : Form
{
  DbContext _db;

  async Task HandleButtonClick()
  {
    var count1 = await _db.MyTable.LoadWith(x => x.Other).Where(x => true).CountAsync(); // Works
    var list1 = await _db.MyTable.LoadWith(x => x.Other).Where(x => true).ToListAsync(); // Works

    var count2 = await _db.MyTable.Where(x => true).LoadWith(x => x.Other).ToCountAsync(); // Works
    var list2 = await _db.MyTable.Where(x => true).LoadWith(x => x.Other).ToListAsync(); // Thread crash
  }
}

```

Double-clicking that button leads to a thread crash in the last call, because of the earlier mentioned observations. It seems the design pattern is not explicitly illegal (as per your clarification), but I can't see the justification why that last line is illegal while the others are not?

**Comment by viceroypenguin** [comment]
> But that click handler is invoked again if you click twice

double-clicking is definitely in the realm of "sharing across multiple threads simultaneously", even if it's all done on the UI thread. in that case, you need to either use a concurrency primitive (`SemaphoreSlim`, etc.) to limit access to the code from multiple concurrently operating handlers, or you need to create a new context for each click/handler. 

The main point is that it is not valid behavior to run two db queries simultaneously on one db context, whether that's because of an `await` situation or because of a multi-thread situation, or whatever else. 

Any reliability or success in doing so is coincidental, and not intentional; and will eventually break.

**Comment by viceroypenguin** [comment]
@cal-tlabwest I agree. #4405 was added yesterday to discuss doing just that.

**Comment by viceroypenguin** [comment]
Closing as `wontfix`, in favor of #4405
